Add `customBooleanField(checkbox)` support

B77Mills opened the pull request:
Add the ability to create & edit boolean fields instead of only allowing select type fields.  

### Create
<img width="1792" alt="Screen Shot 2022-01-03 at 10 52 29 AM" src="https://user-images.githubusercontent.com/3845869/147957519-e4ab78f4-8271-458c-a105-7e4517385a93.png">

### Update
<img width="1792" alt="Screen Shot 2022-01-03 at 10 53 03 AM" src="https://user-images.githubusercontent.com/3845869/147957545-4e1cbaee-b0c8-4107-9fbc-3b9a3b5b50c4.png">

### Update to handel app-user update interface

<img width="884" alt="Screen Shot 2022-01-11 at 7 27 34 AM" src="https://user-images.githubusercontent.com/3845869/148959993-167f5b49-d5d6-447e-b0b2-214f98088014.png">

B77Mills committed:
Add BooleanField to mongo model and schema

B77Mills committed:
Add BooleanField to graph definitions & resolvers

B77Mills committed:
add the buildFragmentTypes Logic

Fix issue with applo-cache-inmememory issue & dynamically create the fragment types

B77Mills committed:
split the edit fields route into multipe

Add support to edit /select/${field_Id} & /boolean/${field_id}

B77Mills committed:
Add boolean field support to the application actions

Add support for create & upadte & find for boolean fields type

B77Mills committed:
cleanup saving of select field

B77Mills committed:
Add ability to create and set field type on create

Allow for create of select & boolean field types.

B77Mills committed:
fix if/else/unless issue

B77Mills committed:
Add userBooleanAnser action

B77Mills committed:
Add CustomBooleanFieldAnswerSchema to app-user schema

B77Mills committed:
Add update action to boolean answers

B77Mills committed:
Add graph definition & resolver for custom boolean fields

B77Mills committed:
add edit route & controller for app user’s boolean custom fields

B77Mills committed:
add template for new custom boolean field routes

B77Mills committed:
Fix lint error

B77Mills committed:
CustomBooleanField clean up & render html label

Upadte how the label is displayed and fix the user boolean ansers actions

B77Mills committed:
Adding this back in.

Should have not been removed, but only rendered on Select type questions.

B77Mills committed:
fix lint error

will have to figure out how to render as html

B77Mills commented:
![](https://github.trello.services/images/mini-trello-icon.png) [How to add checkbox fields to Identity-X](https://trello.com/c/adTAKA4F/258-how-to-add-checkbox-fields-to-identity-x)

zarathustra323 reviewed (commented):
On the custom checkboxes view page for a user, the field name and whether it's active should also be displayed, similar to what's showing on custom select fields:

![image](https://user-images.githubusercontent.com/3289485/148790607-740836c6-09da-42dd-afd5-cd62ccb2fd10.png)

B77Mills committed:
adjust custom boolean card

B77Mills commented:
@zarathustra323 I added the name and Active status to the customBoolean tab on the app-user interface.  
<img width="884" alt="Screen Shot 2022-01-11 at 7 27 34 AM" src="https://user-images.githubusercontent.com/3845869/148960477-96cffaff-1dec-44a1-be5b-e0cd92829e7c.png">

B77Mills committed:
remove extra return

B77Mills committed:
udpate to custom question answer updates

Instead of setting the custom select and boolean fields on app user update to the ones being currently sent(active) to also included previously answered, nonActive, questions.  This prevents lossing legacy data.

zarathustra323 commented on services/application/src/actions/user/update-custom-boolean-answers.js:
```diff
@@ -21,9 +21,24 @@ module.exports = async ({
-    .map(({ fieldId, value }) => ({ _id: fieldId, value }));
+  // get all current answers as object { id, value }
+  const userObj = user.customBooleanFieldAnswers.reduce(
+    (obj, item) => Object.assign(obj, { [item._id]: item.value }), {},
```
This should use rest spread (same as assign but cleaner)

`(obj, item) => ({ ...obj, [item._id]: item.value }), {},`

zarathustra323 commented on services/application/src/actions/user/update-custom-boolean-answers.js:
```diff
@@ -21,9 +21,24 @@ module.exports = async ({
+
+  // get new answers as object { id, value }
+  const newAnswers = answers.reduce(
+    (obj, item) => Object.assign(obj, { [item.fieldId]: item.value }), {},
```
Same here :)

zarathustra323 commented on services/application/src/actions/user/update-custom-select-answers.js:
```diff
@@ -21,9 +21,26 @@ module.exports = async ({
-  const toSet = answers
+  // get all current answers as object { id, value }
+  const userObj = user.customSelectFieldAnswers.reduce(
+    (obj, item) => Object.assign(obj, { [item._id]: item.values }), {},
```
Convert to rest spead

zarathustra323 commented on services/application/src/actions/user/update-custom-select-answers.js:
```diff
@@ -21,9 +21,26 @@ module.exports = async ({
-    .map(({ fieldId, optionIds }) => ({ _id: fieldId, values: optionIds }));
+    .map(({ fieldId, optionIds }) => ({ _id: fieldId, values: optionIds }))
+    .reduce(
+      (obj, item) => Object.assign(obj, { [item._id]: item.values }), {},
```
Convert to rest spread

zarathustra323 commented on services/graphql/src/graphql/definitions/app-user.js:
```diff
@@ -117,6 +121,17 @@ type AppUserRegionalConsentAnswer {
+  field: BooleanField!
+  "Whether the user has answered the question."
+  hasAnswered: Boolean!
+  "The answered field option(s). This will always be an array, even if the field is a single-select only. An empty value signifies a non, or no longer valid, answer. It's up to the implementing components to account for this."
```
Change comment to `The answered value.`

zarathustra323 commented on services/graphql/src/graphql/definitions/app-user.js:
```diff
@@ -311,6 +337,18 @@ input UpdateAppUserMutationInput {
+input UpdateAppUserCustomBooleanAnswersMutationInput {
+  "The user id to update."
+  id: String!
+  "The answers to set/update. An empty array will _unset_ all existing answers. A null value will do nothing."
```
I'm not sure the empty array will unset all answers now, will it?

B77Mills replied:
@zarathustra323  It will not.  On empty or null it will do nothing.  I update the comment to match

zarathustra323 commented on services/graphql/src/graphql/definitions/field.js:
```diff
@@ -106,6 +134,21 @@ type SelectFieldOption {
+  required: Boolean = false
+  "Whether the field is currently active."
+  active: Boolean = true
+  "Whether the select field supports multiple answers."
```
This comment wouldn't apply to a boolean field, as multiple answers are not supported. How is `value` used by the mutation?

zarathustra323 replied:
Actually, it doesn't look like the `value` field is used by the resolver. Remove

B77Mills replied:
@zarathustra323  you are correct and this should now be removed

B77Mills committed:
Use spread instead of Object.assign

B77Mills committed:
remove value on create of custome boolean field

B77Mills committed:
update Descriptions of custom questions

udpate boolean value description & explination of what a empty or null array would do.

An empty or null answers array on CustomBoolean or CustomSelect will update nothing instead or setting to empty [] which would remove previously set answers.
